### PR TITLE
[Merged by Bors] - feat: port/CategoryTheory.Limits.IsLimit

### PR DESCRIPTION
**Pull request by mattrobball**
---

- [x] depends on: #2337 
[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#2337
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#2337~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by mattrobball** [comment]
This should be more reasonable for a git history.

File: Mathlib/CategoryTheory/Limits/IsLimit.lean [hunk 1]
```diff
@@ -0,0 +1,1085 @@
...
+  fac : ∀ (s : Cone F) (j : J), lift s ≫ t.π.app j = s.π.app j := by aesop_cat
+  /-- It is the unique such map to do this -/
+  uniq : ∀ (s : Cone F) (m : s.pt ⟶ t.pt) (_ : ∀ j : J, m ≫ t.π.app j = s.π.app j), m = lift s := by
+    aesop_cat
+#align category_theory.limits.is_limit CategoryTheory.Limits.IsLimit
+#align category_theory.limits.is_limit.fac' CategoryTheory.Limits.IsLimit.fac
+#align category_theory.limits.is_limit.uniq' CategoryTheory.Limits.IsLimit.uniq
+
+-- Porting note: linter claimed it reduced but it did not
+attribute [nolint simpNF] IsLimit.mk.injEq
```
Comment (adamtopaz) [inline Mathlib/CategoryTheory/Limits/IsLimit.lean:73]: It's silly, yes, but the LHS of this lemma does reduce (to `True`) due to the subsingleton instance on line 79. 
I don't know what the right fix is here.

(Similarly for the colimit case.)

File: Mathlib/CategoryTheory/Limits/IsLimit.lean [hunk 2]
```diff
@@ -0,0 +1,1085 @@
...
+variable {t : Cone F}
+
+theorem hom_lift (h : IsLimit t) {W : C} (m : W ⟶ t.pt) :
+    m =
+      h.lift
+        { pt := W
+          π := { app := fun b => m ≫ t.π.app b } } :=
+  h.uniq
+    { pt := W
+      π := { app := fun b => m ≫ t.π.app b } } m fun b => rfl
```
Comment (adamtopaz) [inline Mathlib/CategoryTheory/Limits/IsLimit.lean:236]:
```suggestion
    m = h.lift ⟨W, { app := fun b => m ≫ t.π.app b }⟩ :=
  h.uniq ⟨W, { app := fun b => m ≫ t.π.app b }⟩ m fun b => rfl
```

... and similarly for `hom_desc`.

Not strictly required, but I think it looks nicer.

File: Mathlib/CategoryTheory/Limits/IsLimit.lean [hunk 3]
```diff
@@ -0,0 +1,1085 @@
...
+end Equivalence
+
+/-- The universal property of a limit cone: a map `W ⟶ X` is the same as
+  a cone on `F` with vertex `W`. -/
+def homIso (h : IsLimit t) (W : C) : ULift.{u₁} (W ⟶ t.pt : Type v₃) ≅ (const J).obj W ⟶ F where
+  hom f := (t.extend f.down).π
+  inv π :=
+    ⟨h.lift
+        { pt := W
+          π }⟩
```
Comment (adamtopaz) [inline Mathlib/CategoryTheory/Limits/IsLimit.lean:408]:
```suggestion
  inv π := ⟨h.lift { pt := W, π }⟩
```

File: Mathlib/CategoryTheory/Limits/IsLimit.lean [hunk 4]
```diff
@@ -0,0 +1,1085 @@
...
+    (ofRightAdjoint (Cones.whiskeringEquivalence e).inverse P : _)
+#align category_theory.limits.is_limit.of_whisker_equivalence CategoryTheory.Limits.IsLimit.ofWhiskerEquivalence
+
+/-- Given an equivalence of diagrams `e`, `s` is a limit cone iff `s.whisker e.functor` is.
+-/
+def whiskerEquivalenceEquiv {s : Cone F} (e : K ≌ J) : IsLimit s ≃ IsLimit (s.whisker e.functor) :=
+  ⟨fun h => h.whiskerEquivalence e, ofWhiskerEquivalence e, by aesop_cat, by aesop_cat⟩
+#align category_theory.limits.is_limit.whisker_equivalence_equiv CategoryTheory.Limits.IsLimit.whiskerEquivalenceEquiv
+
+/-- We can prove two cone points `(s : Cone F).pt` and `(t.Cone G).pt` are isomorphic if
```
Comment (adamtopaz) [inline Mathlib/CategoryTheory/Limits/IsLimit.lean:369]:
```suggestion
/-- We can prove two cone points `(s : Cone F).pt` and `(t : Cone G).pt` are isomorphic if
```
... similarly in the cocone section.

File: Mathlib/CategoryTheory/Limits/IsLimit.lean [hunk 5]
```diff
@@ -0,0 +1,1085 @@
...
+      simp
+    inv_hom_id := by
+      apply hom_ext Q
+      aesop_cat }
+#align category_theory.limits.is_limit.cone_points_iso_of_equivalence CategoryTheory.Limits.IsLimit.conePointsIsoOfEquivalence
+
+end Equivalence
+
+/-- The universal property of a limit cone: a map `W ⟶ X` is the same as
+  a cone on `F` with vertex `W`. -/
```
Comment (adamtopaz) [inline Mathlib/CategoryTheory/Limits/IsLimit.lean:402]:
```suggestion
  a cone on `F` with cone point `W`. -/
```
... similarly in a few other places in this file.

File: Mathlib/CategoryTheory/Limits/IsLimit.lean (hunk 1 above)
Comment (mattrobball) [reply Mathlib/CategoryTheory/Limits/IsLimit.lean:73]: `simp` can prove this

File: Mathlib/CategoryTheory/Limits/IsLimit.lean (hunk 2 above)
Comment (mattrobball) [reply Mathlib/CategoryTheory/Limits/IsLimit.lean:236]: ✅

File: Mathlib/CategoryTheory/Limits/IsLimit.lean (hunk 3 above)
Comment (mattrobball) [reply Mathlib/CategoryTheory/Limits/IsLimit.lean:408]: ✅

File: Mathlib/CategoryTheory/Limits/IsLimit.lean (hunk 4 above)
Comment (mattrobball) [reply Mathlib/CategoryTheory/Limits/IsLimit.lean:369]: ✅

File: Mathlib/CategoryTheory/Limits/IsLimit.lean (hunk 5 above)
Comment (mattrobball) [reply Mathlib/CategoryTheory/Limits/IsLimit.lean:402]: I furthered the propaganda.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**